### PR TITLE
#7340  Add vswhere for builds

### PR DESCRIPTION
**Pull request by StingyJack**
## Bug
Fixes: https://github.com/nuget/home/issues/7340
Regression: Yes/No  No
If Regression then when did it last work:   
If Regression then how are we preventing it in future:   

## Fix
Details: uses vswhere to locate msbuild instead of assuming its in the c: programfiles enterprise vs folder

## Testing/Validation
Tests Added: No
Reason for not adding tests:  its in the powershell build script and there isnt a pester project setup for those
Validation done:  This isnt pharmaceutical industry work, so there is not a need for [Validation](https://en.wikipedia.org/wiki/Validation_(drug_manufacture)). If the question being asked is did I bother testing it, then yes "it works on my machine now"

### CODE REVIEW

File: build/common.ps1 [hunk 1]
```diff
@@ -189,12 +189,36 @@ Function Install-NuGet {
...
 }
 
+Function Get-MsBuildPath{
+    [CmdletBinding()]
+    param()
+
+    Push-Location $PSScriptRoot
+
+    try {
+        $path = .\vswhere.exe -latest -products * -requires Microsoft.Component.MSBuild -property installationPath
```
Comment (nkolev92) [inline build/common.ps1]: vswhere is not added part of the PATH when VS is installed. 

You'd need to use the full location of the exe for the script to work universally. 

Not 100% where all this msbuild path is used, but I don't know if it's a good idea to default to the latest version of MSBuild we can find on the machine.

Comment (nkolev92) [reply build/common.ps1]: oh I just noticed you added the binary in repo. That doesn't belong there.

**Review by nkolev92 — COMMENTED** [review]
Can you please remove the sln and csproj changes? 
Furthermore we shouldn't be uploading binaries in the repo unless absolutely necessary.

File: build/common.ps1 (hunk 1 above)
Comment (StingyJack) [reply build/common.ps1]: if it doesnt belong there, _vswhere_ does it need to go?   

That is more or less a copy of the [example](https://github.com/Microsoft/vswhere/wiki/Find-MSBuild), and the problem with the existing script is that its hardcoded to use the `c:\program files(x86)\...` path, and that is not where I have a vs installation at, so the script fails for anyone not using that default location.   `vswhere` is supposed to be the way to locate vs instances since the registry is no longer reliable/used. 

Utility binaries go in repos all the time. Its not evil when its needed. EDIT: Its also how you can solve the problem you raised about which version of MSBuild to use.

Comment (StingyJack) [reply build/common.ps1]: I did add the files to the SLN as those kind of artifacts are part of the solution and the code that must be maintained together  If there is a reason aside from a personal preference why they should not be present, please share. 

VS added the proj file changes. I didn't initiate those. It does that kind of often, and it should stop creating that kind of version control churn. EDIT: What I mean is that I can try to revert it, but VS may just do it again.

Comment (StingyJack) [reply build/common.ps1]: if having vswhere.exe is a significant problem, I've used the VsSetup powershell module to locate MsBuild before. Is installing a supporting module like this generally acceptable or should it be a Save-Module, Import-Module?

Comment (nkolev92) [reply build/common.ps1]: as far as I am aware, vswhere.exe will always be located in "C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe" 
I don't think you can change the location of the installer. 

So you can probably just hardcode that. 

As far as binaries in the repo, that's a no because it drastically increases the size of the repo forever, and it's a good rule of thumb to acquire tools like that rather than package them in. 

Basically cleanup the VS added fluff and just depend on the local version of vswhere and it should be a lot better.

Comment (zivkan) [reply build/common.ps1]: The wiki page that StingyJack linked says
> With Visual Studio 2017 Update 2 or newer installed, you can find vswhere at %ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe

VS release notes say that 15.0.2 was released in March 2017, and 15.2 was released in May 2017, so either way anyone who has updated/installed in the last 17 months should have it at that well defined location.

The linked docs also say that vswhere is available as a nuget package, so the powershell script could check for vswhere at the well known location, and if it's it's not there, then download it with nuget.exe. Both configure.ps1 and build.ps1 run install-nuget, so as long as Get-MsBuildPath is called after Install-NuGet, we know nuget.exe exists at .nuget\nuget.exe. Or just add vswhere to .nuget\packages.config and then we know that vswhere will always be installed to the packages folder and don't even try the VS Installer path.

Comment (StingyJack) [reply build/common.ps1]: > VS release notes say that 15.0.2 was released in March 2017, and 15.2 was released in May 2017, so either way anyone who has updated/installed in the last 17 months should have it at that well defined location.

Based on my observations and experiences, this makes an assumption that VS users regularly take the updates issued. If I dont see anything useful in the release notes, I wont update sometimes for months. There are a few people that I work with and know that have not taken a VS 15 update for a much longer time primarily because they do not want to risk destabilizing their development environment but also for other reasons. With the recent announcement of 15.9 becoming "service pack 1", and all the interim versions between it and RTM going "unsupported", they will probably soon have to take an even larger risk. 

This isn't a dispute of your suggestions as realistically none of them are likely to ever try to run this script, or even be curious about the repo at all (or even know GH exists), but its important to remember these are your "moderately frequent adopters" and they do exist in the userbase, and its not abnormal for them to go 1-3 years between taking even minor updates. For them, this speed "feels" normal and sometimes even too fast relative to the infrequent adopters they tend to work with (end customers in healthcare and manufacturing will run the same software for 10-20 years or more). 



> Basically cleanup the VS added fluff and just depend on the local version of vswhere and it should be a lot better.

I cant guarantee that VS wont re-add the fluff,  but I'll try to get to the suggested updates over the weekend.

**Comment by StingyJack** [comment]
@nkolev92 - remade the commit per suggestion. My VS install path is E:\Vs2017, and the `build.ps1` runs without error when executed without params and with -SkipUnitTest. 

There are quite a few warnings that are shown, but I think those are pre-existing.

**Review by nkolev92 — APPROVED** [review]
LGTM. 

Thanks for the contribution. 

Kicked off the tests.

File: build/common.ps1 [hunk 2]
```diff
@@ -189,12 +190,34 @@ Function Install-NuGet {
...
+    param()
+
+    if (-Not (Test-Path $VsWhereExe))
+    {
+        throw "Expected VsWhere.exe at '$VsWhereExe' but it was not present. This is needed to locate MSBuild.exe"
+    }
+
+    $path = & $VsWhereExe -latest -products * -requires Microsoft.Component.MSBuild -property installationPath
+    if ($path) {
+        $path = join-path $path 'MSBuild\15.0\Bin\MSBuild.exe'
```
Comment (jainaashish) [inline build/common.ps1:204]: Since MSBuild is moving to `current` folder going forward so we should update this code accordingly as well. Look at https://github.com/NuGet/NuGet.Client/pull/2487

**Review by jainaashish — CHANGES_REQUESTED** [review]

File: build/common.ps1 (hunk 2 above)
Comment (nkolev92) [reply build/common.ps1:204]: @jainaashish 

I think that change belongs in a different PR. 

VSWhere ships everywhere, so the fallback will likely never be used.

Comment (jainaashish) [reply build/common.ps1:204]: its about the getting msbuild.exe path not vswhere and for now it will use fallback (like `msbuild\15.0\bin...) instead of `msbuild\current\bin...

Comment (nkolev92) [reply build/common.ps1:204]: oh my bad, didn't look at the change again. 

I still think the point stands. 
We need to test for both and they can/should be done in a different PR.

File: build/common.ps1 [hunk 3]
```diff
@@ -189,12 +190,34 @@ Function Install-NuGet {
...
+
 Function Install-DotnetCLI {
     [CmdletBinding()]
     param(
         [switch]$Force
     )
-    $msbuildExe = 'C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\bin\msbuild.exe'
+
+
+    $msbuildExe = Get-MsBuildPath
```
Comment (nkolev92) [inline build/common.ps1:220]: Just looking at the scripts for an unrelated task. 

Doesn't using Get-MSBuildExe fix it already?

File: build/common.ps1 (hunk 2 above)
Comment (heaths) [reply build/common.ps1:204]: Actually, we didn't start shipping vswhere.exe in the Visual Studio Installer until ~Visual Studio 2017 version 15.2, so there is a small window where it might not be available on some customers' machines. I recommend keeping the fallback.

File: build/common.ps1 (hunk 3 above)
Comment (StingyJack) [reply build/common.ps1:220]: im not following what you are asking about this

File: build/common.ps1 (hunk 2 above)
Comment (StingyJack) [reply build/common.ps1:204]: @heaths - Are there lots of 15.0's out there? The window may be larger with the announcement of 15.9 becoming SP1 and everything between it and 15.0 becoming unsupported.

Comment (heaths) [reply build/common.ps1:204]: I can't give exact numbers, but enough you might still want to consider it.

And, for the record, 15.9 is not becoming "SP1" - just the long-term servicing (LTS) baseline.

File: build/common.ps1 (hunk 3 above)
Comment (nkolev92) [reply build/common.ps1:220]: Instead of using the VSWhere logic to find MSBUild, does simply using that function solve the problem? 
Get-MSBuildExe is already something that exists and a quick look suggests that it might handle the VS on a different disk scenario.

Comment (StingyJack) [reply build/common.ps1:220]: No, that thing that assumes its on the drive with program files x86.  I dont even have the VS15COMNTOOLS environment variable its talking about, just the ones for 12.0 and 14,.0,

For completeness it should probably be changed also.

**Comment by heaths** [comment]
An alternative that also fixes many other locations where MSBuild was being "looked up" (rather, assumed) differently. Also fixes another issue: https://github.com/NuGet/NuGet.Client/pull/2526

**Comment by jainaashish** [comment]
Thanks @StingyJack for taking time to create this PR. But I'm closing this in favor of https://github.com/NuGet/NuGet.Client/pull/2526 

Please continue to contribute to NuGet to improvise.

**Comment by StingyJack** [comment]
@jainaashish  - thats fine with me. I dont mind sharing something I've had to fix, but im not on the payroll and cant dedicate that much time to improvising.
